whylog/config/abstract_config.py

```diff
@@ -166,22 +165,69 @@ def _create_steps_in_investigation(self, concatenated_parsers, suspected_rules,
-                    'right_bound': datetime(2016, 4, 12, 23, 54, 33)
-                }
-            },
-            'apache': {
+    @classmethod
+    def _get_search_ranges(cls, suspected_rules, effect_clues):
+        """
+        search_range is a dictionary with two keys: InvesitgationStep.LEFT_BOUND and
+        InvesitgationStep.RIGHT_BOUND. Values in this dict are concrete values of type the same as type of some
+        primary key. Both values taken together represent some interval.
+        Sample search range:
+            {
+                InvesitgationStep.LEFT_BOUND: datetime(2016, 5, 29, 12, 33, 0),
+                InvesitgationStep.RIGHT_BOUND: datetime(2016, 5, 29, 12, 33, 30)
+            }
+        Log type's search ranges in rule context is a dictionary with search_range for every type of primary key of
+        rule's parsers that belong to this log type.
+        Sample log type's search ranges:
+            {
                 'date': {
-                    'left_bound': datetime(2016, 4, 12, 23, 54, 33),
-                    'right_bound': datetime(2016, 4, 12, 23, 54, 43)
+                    InvesitgationStep.LEFT_BOUND: datetime(2016, 5, 29, 12, 33, 0),
+                    InvesitgationStep.RIGHT_BOUND: datetime(2016, 5, 29, 12, 33, 30)
                 }
             }
-        }
+        Rule's search_ranges is sum of all log type's search ranges, where rule's parsers belong to these log types.
+        Sample rule's search ranges:
+            Sample search_ranges: {
+                'apache': {
+                    'date': {
+                        InvesitgationStep.LEFT_BOUND: datetime(2016, 5, 29, 12, 33, 0),
+                        InvesitgationStep.RIGHT_BOUND: datetime(2016, 5, 29, 12, 33, 30)
+                    }
+                },
+                'database': {
+                    'date': {
+                        InvesitgationStep.LEFT_BOUND: datetime(2016, 5, 29, 12, 32, 0),
+                        InvesitgationStep.RIGHT_BOUND: datetime(2016, 5, 29, 12, 33, 20)
+                    }
+                }
+             }
+        This method sums all rule's search ranges from every rule in suspected rules.
+        """
+        # This implementation assumes that all primary key groups is a one element list
+        # TODO implementation for longer primary key groups
+        search_ranges = {}
+        for rule in suspected_rules:
+            rule_search_ranges = rule.get_search_ranges(effect_clues)
+            for log_type_name, log_type_ranges in six.iteritems(rule_search_ranges):
+                if log_type_name not in search_ranges:
+                    search_ranges[log_type_name] = log_type_ranges
+                    continue
+                log_type_search_range = search_ranges[log_type_name]
+                for key_type, type_range in six.iteritems(rule_search_ranges[log_type_name]):
+                    if key_type not in log_type_search_range:
```

Comment:
Please get it by get() and then just check if it is None instead of double lookup (a few lines below)